colexec: clean up remapping of indexed vars for filter expressions

yuzefovich opened the pull request:
**colexec: fix handling of filters by joiners**

Previously, in order to handle ON expression of the joiners, we would
modify the expression itself (i.e. we would remap the IndexedVars inside
of the expression). However, this approach is error-prone and doesn't
work in all cases (consider an example when we have a filter like
"@1 = 'abc@2def'" - @2 is not an ordinal, but it would get treated as an
IndexedVar with index 1). Now we have enhanced the IndexedVarHelper to
handle the remapping internally when the IndexedVar is being bound to
a container. This way no modifications to the actual expressions are
needed, and such approach should work in all cases.

Fixes: #41407.
Fixes: #41944.
Fixes: #42100.

**sem/tree: remove unused field from IndexedVar**

This commit removes 'bindInPlace' field from IndexedVar struct since it
is not being used anywhere.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42198)
<!-- Reviewable:end -->

yuzefovich committed:
colexec: fix handling of filters by joiners

Previously, in order to handle ON expression of the joiners, we would
modify the expression itself (i.e. we would remap the IndexedVars inside
of the expression). However, this approach is error-prone and doesn't
work in all cases (consider an example when we have a filter like
"@1 = 'abc@2def'" - @2 is not an ordinal, but it would get treated as an
IndexedVar with index 1). Now we have enhanced the IndexedVarHelper to
handle the remapping internally when the IndexedVar is being bound to
a container. This way no modifications to the actual expressions are
needed, and such approach should work in all cases.

Release note: None

yuzefovich committed:
sem/tree: remove unused field from IndexedVar

This commit removes 'bindInPlace' field from IndexedVar struct since it
is not being used anywhere.

Release note: None

asubiotto reviewed (changes requested):
> Reviewed 5 of 5 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42198)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colexec/execplan.go, line 780 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42198#-Lt0GYAm-Xe8V1Due1IT:-Lt0GYAm-Xe8V1Due1IU:ba8u6jn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0b03ff4bc148e205be1d9863b07f4b3e6a8380/pkg/sql/colexec/execplan.go#L780)):*
> ```Go
> 	)
> 	if err != nil {
> 		execerror.VectorizedInternalPanic(
> ```

I don't think we should panic during plan time
___
*[pkg/sql/colexec/expr.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42198#-Lt0HeNlBPwtLMyX0Glw:-Lt0HeNlBPwtLMyX0Glx:b-b14q61) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0b03ff4bc148e205be1d9863b07f4b3e6a8380/pkg/sql/colexec/expr.go#L30)):*
> ```Go
> 		exprToWalk = expr.LocalExpr
> 	} else {
> 		e, err := parser.ParseExpr(expr.Expr)
> ```

Should we update `expr.LocalExpr` with the parsed expression?
___
*[pkg/sql/logictest/testdata/logic_test/exec_hash_join, line 141 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42198#-Lt0Ht_y4plPsIRFKmlQ:-Lt0Ht_y4plPsIRFKmlR:b-fd548x) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0b03ff4bc148e205be1d9863b07f4b3e6a8380/pkg/sql/logictest/testdata/logic_test/exec_hash_join#L141)):*
> ```
> 
> query T
> EXPLAIN (VEC)
> ```

Should we also execute this query?
___
*[pkg/sql/sem/tree/indexed_vars.go, line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42198#-Lt0IE3X0lj3FIoxtTYw:-Lt0IE3X0lj3FIoxtTYx:b-hfd1uz) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0b03ff4bc148e205be1d9863b07f4b3e6a8380/pkg/sql/sem/tree/indexed_vars.go#L137)):*
> ```Go
> // getActualIndex is a helper function that returns an "actual index" of the
> // IndexedVar.
> func (h *IndexedVarHelper) getActualIndex(ivar *IndexedVar) int {
> ```

Could we call this just `getIndex`?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42198)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colexec/execplan.go, line 780 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42198#-Lt0GYAm-Xe8V1Due1IT:-Lt1DMvGCnE4pT4FjrzE:bwarzoy) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0b03ff4bc148e205be1d9863b07f4b3e6a8380/pkg/sql/colexec/execplan.go#L780)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't think we should panic during plan time
</blockquote></details>

Done. I was being lazy - we would have caught the panic within `SupportsVectorized` check.
___
*[pkg/sql/colexec/expr.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42198#-Lt0HeNlBPwtLMyX0Glw:-Lt1EOYd74Yarz6fcioG:bnnojjs) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0b03ff4bc148e205be1d9863b07f4b3e6a8380/pkg/sql/colexec/expr.go#L30)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Should we update `expr.LocalExpr` with the parsed expression?
</blockquote></details>

I briefly thought about it and decided not to for a couple of reasons:
1. `LocalExpr` is `tree.TypedExpr` whereas after parsing we only have `tree.Expr`, so we would need to add extra logic to simulate what `processExpression` (from `execinfra/expr.go`) does. We would need to do the type checking and binding the IndexedVars to a container. We might not have all the needed information at this point, but also we might end up binding it to a wrong container.
2. I modified this `execinfrapb.Expression` in-place at some point, and it bit me back. I think it is less error-prone to not update anything in-place (unless you fully understand the consequences - which I don't in this case).

True, we will have to parse the filter expression on each remote node - I think - three times (for `SupportsVectorized` check, for actual vectorized flow setup, and for actually executing the filter), but at least it should be correct, and the performance hit will be probably minor in comparison to other vectorized things.
___
*[pkg/sql/logictest/testdata/logic_test/exec_hash_join, line 141 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42198#-Lt0Ht_y4plPsIRFKmlQ:-Lt1G5h87PILsiYOKVc2:bk3ga8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0b03ff4bc148e205be1d9863b07f4b3e6a8380/pkg/sql/logictest/testdata/logic_test/exec_hash_join#L141)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Should we also execute this query?
</blockquote></details>

I think doing `EXPLAIN (VEC)` and executing the query is redundant in this case, so I replaced the explain with the execution.
___
*[pkg/sql/sem/tree/indexed_vars.go, line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42198#-Lt0IE3X0lj3FIoxtTYw:-Lt1GJmI5_a6WclzQo2S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f0b03ff4bc148e205be1d9863b07f4b3e6a8380/pkg/sql/sem/tree/indexed_vars.go#L137)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Could we call this just `getIndex`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r3, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42198#-:-Lt1HaNQEdWC8Y3GJEKx:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
